### DISCUSSION TITLE
PDF tags support

### DISCUSSION THREAD

**Discussion by hellpanderrr** [pymupdf/PyMuPDF#2764] [Q&A] [answered]
Are [PDF tags](https://accessible-pdf.info/en/basics/general/overview-of-the-pdf-tags/) supported in pymupdf? If not, is it possible to get them by xref?

 Something like:
`doc.xref_object(doc.pdf_catalog())`
returns
`<<\n  /Lang (FR-FR)\n  /MarkInfo <<\n    /Marked true\n  >>\n  /Metadata 392 0 R\n  /OCProperties <<\n    /D <<\n      /Order [ 6082 0 R ]\n      /RBGroups [ ]\n    >>\n    /OCGs [ 6082 0 R ]\n  >>\n  /Outlines 480 0 R\n  /PageLayout /SinglePage\n  /Pages 6050 0 R\n  /StructTreeRoot 623 0 R\n  /Type /Catalog\n>>`
Then using **/StructTreeRoot 623** 
`doc.xref_object(623)`
I get 
`<<\n  /K [ 6047 0 R 5790 0 R 5446 0 R 5233 0 R 5220 0 R 5062 0 R 4982 0 R\n      4972 0 R 4878 0 R 4789 0 R 4711 0 R 4643 0 R 4530 0 R 4441 0 R\n      4365 0 R 4291 0 R 4219 0 R 4158 0 R 4070 0 R 3982 0 R 3904 0 R\n      3834 0 R 3762 0 R 3690 0 R 3645 0 R 3564 0 R 3494 0 R 3414 0 R\n      3342 0 R 3281 0 R 3211 0 R 3139 0 R 3065 0 R 2993 0 R 2923 0 R\n      2851 0 R 2793 0 R 2733 0 R 2641 0 R 2575 0 R 2511 0 R 2447 0 R\n      2383 0 R 2317 0 R 2259 0 R 2201 0 R 2143 0 R 2077 0 R 2009 0 R\n      1945 0 R 1877 0 R 1867 0 R 1788 0 R 1706 0 R 1634 0 R 1566 0 R\n      1502 0 R 1436 0 R 1426 0 R 1358 0 R 1348 0 R 1238 0 R 1228 0 R\n      1123 0 R 1059 0 R 1051 0 R 970 0 R 898 0 R 836 0 R 771 0 R\n      721 0 R ]\n  /Namespaces [ 691 0 R ]\n  /ParentTree 624 0 R\n  /ParentTreeNextKey 188\n  /RoleMap 625 0 R\n  /Type /StructTreeRoot\n>>`
Looks like this gives me the actual tag item:
```
>>> doc.xref_object(6047)
<<\n  /K [ 6037 0 R 6038 0 R 6046 0 R ]\n  /Lang (EN-US)\n  /P 623 0 R\n  /S /Slide\n>>
```

Am I on the right track here or is there any other way?

**Comment by JorjMcKie** [comment, marked as answer]
Tags are not currently supported.
But you are on the right track:
```python
In [1]: import fitz
In [2]: doc=fitz.open("adobe.pdf")
In [3]: cat=doc.pdf_catalog()
In [4]: doc.xref_get_key(cat, "StructTreeRoot")
Out[4]: ('xref', '109960 0 R')
In [5]:
```
Usually this is an xref, but could be a dict. So check item[0] of the return code.
If an xref, follow that path further ...
